File: 264_nth_ugly_number/nth_ugly_number.cpp
/*
 * @Date: 2021-04-11 09:55:25
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-11 10:03:54
 */

#include <cassert>
#include <vector>

using namespace std;

int nth_ugly_number(int n) {
  vector<int> ugly(n, 1), idx(3, 0);
  for (int i = 1; i < n; ++i) {
    int a = ugly[idx[0]] * 2, b = ugly[idx[1]] * 3, c = ugly[idx[2]] * 5;
    int next = std::min(a, std::min(b, c));
    if (next == a) {
      ++idx[0];
    }
    if (next == b) {
      ++idx[1];
    }
    if (next == c) {
      ++idx[2];
    }
    ugly[i] = next;
  }
  return ugly.back();
}

int main() {
  assert(nth_ugly_number(10) == 12);
  assert(nth_ugly_number(1) == 1);
  return 0;
}

File: rust/src/264_nth_ugly_number/nth_ugly_number.rs
/*
 * @Date: 2021-04-11 09:55:31
 * @Author: Mengsen Wang
 * @LastEditors: Mengsen Wang
 * @LastEditTime: 2021-04-11 10:11:56
 */

fn nth_ugly_number(n: i32) -> i32 {
    let n = n as usize;
    let mut ugly = vec![1; n];
    let mut idx = vec![0; 3];
    for i in 1..n {
        let a = ugly[idx[0]] * 2;
        let b = ugly[idx[1]] * 3;
        let c = ugly[idx[2]] * 5;
        let next = std::cmp::min(c, std::cmp::min(a, b));
        if next == a {
            idx[0] += 1;
        }
        if next == b {
            idx[1] += 1;
        }
        if next == c {
            idx[2] += 1;
        }
        ugly[i] = next;
    }

    ugly.pop().unwrap()
}

fn main() {
    assert_eq!(nth_ugly_number(10), 12);
    assert_eq!(nth_ugly_number(1), 1);
}
